teps/0004-tekton-integrations-annotations-statuses.md

```diff
@@ -0,0 +1,689 @@
+  debugging.
+- Minimize impact on pipeline/task execution spec.
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- While GitHub integrations will often be used as an example, I want to avoid
+  going into too much deep details into how a specific GitHub integration will
+  be configured / what data will be rendered / what APIs will be used / etc,
+  unless it has direct implications on integrations as a whole.
+
+## Requirements
+
+<!--
+List the requirements for this TEP.
+-->
+
+1. Integrations should be optional from core Tekton Pipelines functionality.
+2. Users should be able to pick and choose which integrations they have
+   installed in their cluster.
+3. Integration data should be stored as close to the PipelineRun as possible
+   (ideally alongside).
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Integration Annotations
+
+[Annotations](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/)
+of the format `<name>.integrations.tekton.dev/<value>` to will be used to
+annotate PipelineRun and TaskRun objects with integration metadata. This will
+not be strictly enforced, in order to allow interoperability with other tools /
+annotation namespaces, but we should recommend this as a default by convention.
+
+This fits into the annotations model - from the Kubernetes docs:
+
+> Fields managed by a declarative configuration layer. Attaching these fields as
+> annotations distinguishes them from default values set by clients or servers,
+> and from auto-generated fields and fields set by auto-sizing or auto-scaling
+> systems.
+>
+> Build, release, or image information like timestamps, release IDs, git branch,
+> PR numbers, image hashes, and registry address.
+>
+> User or tool/system provenance information, such as URLs of related objects
+> from other ecosystem components.
+
+This allows for integrations to annotate arbitrary data without needing to
+modify the Pipeline execution spec, with the expectation that another reconciler
+/ tool will notice and act on these annotations.
+
+**Examples:**
+
+- GitHub
+  ```
+  github.integrations.tekton.dev/owner: wlynch
+  github.integrations.tekton.dev/repo: test
+  github.integrations.tekton.dev/sha: 4d2f673c600a0f08c3f6737bc3f5c6a6a15d62e6
+  ```
+- Webhook
+  ```
+  webhook.integrations.tekton.dev/url: https://example.com:8080
+  ```
+- Slack
+  ```
+  slack.integrations.tekton.dev/url: https://tektoncd.slack.com
+  slack.integrations.tekton.dev/channel: build-cop
+  ```
```

Comment:
Who is responsible for adding those ? Are those annotation provided by the user (or something like `triggers`, or both) or by the integration, *or by both* ?

Reply:
Initially, my expectation is the integration would provide the annotation it supports in its documentation, then users would set these fields in their Runs (filling them in with params / TriggerTemplates).

Eventually, I'd like to get to a point where Triggers can do this automatically for users. Initially via ObjectMeta templating, but eventually via interceptors (or possibly some other mechanism) that can also be provided by an integration. e.g. If we know that a Trigger is for GitHub events, and we can filter for these types of events, we should be able to provide the relevant annotations to configure related integration functionality automatically. If users want to configure one-off builds for testing, they can still set the annotations manually.

This also designed in a way that other non-Tekton Trigger mechanisms could also leverage this if they wanted, or we can even mix and match mechanisms to trigger builds and related integrations (e.g. want to build a slack bot that retests pull requests? just set these annotations and you're good to go).

I think the exact details of how Triggers implements these features is a bit hand-wavy at the moment since a lot of it is work actively being discussed / designed, so I didn't want to go too deep in this proposal. The main goal here is to carve out the areas within the Runs API to define this data.